<a href="https://colab.research.google.com/github/el-grudge/mleng-zoomcamp/blob/main/week_8/homework_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
!wget https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip; unzip -q data.zip

--2023-11-18 02:18:01--  https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/e6c56cb7-dce1-463f-865b-01e913c38485?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231118%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231118T021801Z&X-Amz-Expires=300&X-Amz-Signature=8bd94c09e8f6f1af923b04668a573545e312d0b7c58f21928047dabf9217a4b3&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=405934815&response-content-disposition=attachment%3B%20filename%3Ddata.zip&response-content-type=application%2Foctet-stream [following]
--2023-11-18 02:18:01--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/e6c56cb7-dce1-463f-865b-01e913c38485?X-Amz-Algor

In [1]:
import numpy as np
import matplotlib.pyplot as plt

from tensorflow.keras import layers
from tensorflow.keras import models
from tensorflow.keras import optimizers
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.preprocessing import image

In [2]:
# model parameters
filter_count = 32
filter_size = (3,3)
input_shape = (150,150,3)
pooling_size = (2,2)

In [3]:
# model definition
model = models.Sequential()
model.add(layers.Conv2D(filter_count, filter_size, activation='relu', input_shape=input_shape))
model.add(layers.MaxPooling2D(pooling_size))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [4]:
# compile model
model.compile(loss='binary_crossentropy',
             optimizer=optimizers.SGD(learning_rate=0.002, momentum=0.8),
             metrics=['acc'])

#### Question 1

Since we have a binary classification problem, what is the best loss function for us?

* `mean squared error`
* `binary crossentropy`    
* `categorical crossentropy`    
* `cosine similarity`

Answer: `binary crossentropy`

In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 74, 74, 32)        0         
 D)                                                              
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11215873 (42.79 MB)
Trainable params: 11215873 (42.79 MB)
Non-trainable params: 0 (0.00 Byte)
______________

#### Question 2

What's the number of parameters in the convolutional layer of our model? You can use the summary method for that.

Answer: 11214912

In [7]:
# define data generators
train_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/train',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 3677 images belonging to 2 classes.


In [8]:
val_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/test',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 918 images belonging to 2 classes.


In [9]:
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=val_generator
)

Epoch 1/10
184/184 [==============================] - 18s 46ms/step - loss: 0.6915 - acc: 0.5393 - val_loss: 0.6708 - val_acc: 0.5784
Epoch 2/10
184/184 [==============================] - 7s 37ms/step - loss: 0.6686 - acc: 0.5801 - val_loss: 0.6740 - val_acc: 0.5806
Epoch 3/10
184/184 [==============================] - 8s 46ms/step - loss: 0.6332 - acc: 0.6348 - val_loss: 0.5862 - val_acc: 0.6961
Epoch 4/10
184/184 [==============================] - 7s 37ms/step - loss: 0.5770 - acc: 0.7117 - val_loss: 0.5537 - val_acc: 0.7298
Epoch 5/10
184/184 [==============================] - 8s 44ms/step - loss: 0.5544 - acc: 0.7196 - val_loss: 0.5550 - val_acc: 0.6972
Epoch 6/10
184/184 [==============================] - 7s 37ms/step - loss: 0.5169 - acc: 0.7599 - val_loss: 0.5545 - val_acc: 0.7211
Epoch 7/10
184/184 [==============================] - 7s 36ms/step - loss: 0.4980 - acc: 0.7764 - val_loss: 0.5761 - val_acc: 0.6961
Epoch 8/10
184/184 [==============================] - 8s 43ms/step -

#### Question 3

What is the median of training accuracy for all the epochs for this model?

* 0.20
* 0.40
* 0.60
* 0.80

Answer: 0.80

In [10]:
np.median(history.history['acc'])

0.73973348736763

Question 4

What is the standard deviation of training loss for all the epochs for this model?

* 0.031
* 0.061
* 0.091
* 0.131

Answer: 0.091

In [11]:
np.std(history.history['loss'])

0.08371188248319235

In [12]:
train_generator = ImageDataGenerator(
    rescale=1./255,
    rotation_range=50,
    width_shift_range=0.1,
    height_shift_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest').flow_from_directory(
        './data/train',
        target_size=(150,150),
        batch_size=20,
        class_mode='binary',
        shuffle=True)

Found 3677 images belonging to 2 classes.


In [13]:
val_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/test',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 918 images belonging to 2 classes.


In [14]:
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=val_generator
)

Epoch 1/10
184/184 [==============================] - 27s 144ms/step - loss: 0.5089 - acc: 0.7604 - val_loss: 0.5686 - val_acc: 0.7070
Epoch 2/10
184/184 [==============================] - 26s 141ms/step - loss: 0.4963 - acc: 0.7699 - val_loss: 0.4830 - val_acc: 0.7800
Epoch 3/10
184/184 [==============================] - 25s 137ms/step - loss: 0.4985 - acc: 0.7639 - val_loss: 0.5195 - val_acc: 0.7527
Epoch 4/10
184/184 [==============================] - 27s 144ms/step - loss: 0.4860 - acc: 0.7794 - val_loss: 0.5271 - val_acc: 0.7538
Epoch 5/10
184/184 [==============================] - 27s 147ms/step - loss: 0.4813 - acc: 0.7792 - val_loss: 0.4599 - val_acc: 0.7810
Epoch 6/10
184/184 [==============================] - 27s 147ms/step - loss: 0.4737 - acc: 0.7835 - val_loss: 0.5001 - val_acc: 0.7582
Epoch 7/10
184/184 [==============================] - 26s 143ms/step - loss: 0.4626 - acc: 0.7868 - val_loss: 0.4868 - val_acc: 0.7810
Epoch 8/10
184/184 [==============================] - 2

Question 5

Let's train our model for 10 more epochs using the same code as previously.

    Note: make sure you don't re-create the model - we want to continue training the model we already started training.

What is the mean of test loss for all the epochs for the model trained with augmentations?


* 0.18
* 0.48
* 0.78
* 0.108

Answer: 0.48

In [17]:
np.mean(history.history['val_loss'])

0.49224182665348054

Question 6

What's the average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

* 0.38
* 0.58
* 0.78
* 0.98

Answer: 0.78

In [16]:
np.mean(history.history['val_acc'][5:10])

0.7821350693702698